mender-update/standalone/standalone.cpp

```diff
@@ -417,19 +422,75 @@ ResultAndError DoInstallStates(
 		return {Result::FailedNothingDone, payload.error()};
 	}
 
+	auto default_paths {main_context.GetConfig().paths};
+	events::EventLoop loop;
+	auto script_runner {executor::ScriptRunner(
+		loop,
+		chrono::seconds {main_context.GetConfig().state_script_timeout_seconds},
+		default_paths.GetArtScriptsPath(),
+		default_paths.GetRootfsScriptsPath())};
+
+
+	// Download Enter
+	{
+		auto err = script_runner.RunScripts(executor::State::Download, executor::Action::Enter);
+		if (err != error::NoError) {
+			auto leave_error {
+				script_runner.RunScripts(executor::State::Download, executor::Action::Error)};
+			return {Result::FailedNothingDone, leave_error.FollowedBy(err)};
+		}
+	}
 	auto err = update_module.Download(payload.value());
 	if (err != error::NoError) {
 		err = err.FollowedBy(update_module.Cleanup());
 		err = err.FollowedBy(RemoveStateData(main_context.GetMenderStoreDB()));
 		return {Result::FailedNothingDone, err};
 	}
 
+	// Download Leave
+	{
+		auto err = script_runner.RunScripts(executor::State::Download, executor::Action::Leave);
+		if (err != error::NoError) {
+			auto leave_error {
+				script_runner.RunScripts(executor::State::Download, executor::Action::Error)};
+			return {Result::FailedNothingDone, leave_error.FollowedBy(err)};
+		}
+	}
+
+
+	// Install Enter
+	{
+		auto err =
+			script_runner.RunScripts(executor::State::ArtifactInstall, executor::Action::Enter);
+		if (err != error::NoError) {
+			auto install_leave_error {script_runner.RunScripts(
+				executor::State::ArtifactInstall, executor::Action::Error)};
+			return {Result::FailedNothingDone, err.FollowedBy(install_leave_error)};
```

Comment:
I tried reading up on this, but didn't find anything on it. Do we document this anywhere?

Reply:
Well, sort of, but it's kind of documented in reverse. If you look at [the `Download` state from the update module docs](https://github.com/mendersoftware/mender/blob/feature-c%2B%2B-client/Documentation/update-modules-v3-file-api.md#download-state), it says that in this state you should not install or activate anything because it is insecure. This then implies that `ArtifactInstall` is the first state where you can start changing things.

`ArtifactInstall_Enter` are also the first set of scripts which come from the artifact, and not from the filesystem, and hence it also naturally leads to these scripts being used to change the system.